### ISSUE TITLE
Compatilibity with CuDNN R5

### ISSUE THREAD

**Issue by JoostvDoorn** [Element-Research/rnn#239]
@nicholas-leonard Do you have any thoughts on compatibility with CuDNN R5, and possible future roadmap of this package towards maturity? Possibly the modules that have been implemented in soumith/cudnn.torch would be helpful here as well for performance reasons.

**Comment by nicholas-leonard** [comment]
@JoostvDoorn So torch supports this via R5 branch of the cudnn repository : https://github.com/soumith/cudnn.torch/tree/R5 . The rnn package can integrate with these but will not be automatically converted. Similarly to SeqLSTM, they expect tensors of size seqlen x batchsize x inputsize as input. I have already made changes to make integration easier by making Sequencer support Tensors as input as well : https://github.com/Element-Research/rnn/pull/201/commits/8ac27073a874f35496ef0d17695067e9c45b5ad8 . 

What are compatibility would you see fitting?

**Comment by JoostvDoorn** [comment]
I was thinking about automatic converison. In general I think there will be a demand for a module that is compatible with the cudnn modules for RNN/LSTM/GRU and that can then be converted using `cudnn.convert`. Not sure though if this package is the right place to discuss this, but since this is probably the most actively maintained RNN package for torch it seems logical this one would support this eventually. As you state SeqLSTM having the same structure should make this theoretically possible.

**Comment by nicholas-leonard** [comment]
@JoostvDoorn Yes we should definitely at least make SeqLSTM support conversion via cudnn convert. Lets keep this ticket open until it gets done.

**Comment by JoostvDoorn** [comment]
@nicholas-leonard I have some basic conversion code ready for SeqLSTM, here:
https://gist.github.com/JoostvDoorn/bd43ce85a101c0aeee03f734ead688fb

I am mainly wondering how we could include this within cudnn.convert.

**Comment by nicholas-leonard** [comment]
You can always send a PR to rnn so that we can do something like this: 
```lua
require 'cudnn'
require 'rnn'

-- define model
lstm = nn.Sequential()
   :add(...
   :add(nn.SeqLSTM())
    ..)

-- convert to cudnn where possible
culstm = lstm:cuconvert()
```

Not sure if this use-case is simple enough to use and to code. What do you think?

**Comment by hashbangCoder** [comment]
Sorry to but in but is it currently possible to `cudnn.convert` any module? Say a `Sequencer(FastLSTM)` ? 

I have a model which wraps a sequencer around LookupTableZeroMAsk, FastLSTM, Linear (and few others like Dropout) but cudnn.convert only converts the ReLU and ignores even the Linear? 

Let me know if I should move to a new issue thread for this.

**Comment by JoostvDoorn** [comment]
Note that there is no cudnn implementation of nn.Linear. Other than that you are right, and `nn.Sequencer(nn.Tanh)` is not converted to `nn.Sequencer(cudnn.Tanh)`. cudnn.convert uses nn.Module.replace, which recursively converts the children based on `self.modules`, not sure where this goes wrong... @hashbangCoder Can you open a new issue to discuss using SeqLSTM? You don't need to use a Sequencer for the code above.

@nicholas-leonard Yes, I guess that would be the easiest approach.

**Comment by hashbangCoder** [comment]
@JoostvDoorn Yes, I realized that soon after i posted the comment and updated my comment :D
>Can you open a new issue to discuss using SeqLSTM? 

I'm actually interested in the FastLSTM. But if you want me to move this to another thread, sure.

> You don't need to use a Sequencer for the code above.

Could you please elaborate a bit? My understanding is that SeqLSTM is equivalent of (but faster than) `Sequencer(FastLSTM)`. _However_ I want to use something more like this `Sequencer(Sequential():add(Lookup):add(FastLSTM):add(Linear):add(Dropout))` for which I cannot use `SeqLSTM` . I want params to be shared across all layers for all time-steps hence the Sequencer.

It says in the README : 

> Note that a SeqLSTM cannot replace FastLSTM in code that decorates it with a AbstractSequencer or Recursor as this would be equivalent to Sequencer(Sequencer(FastLSTM)). You have been warned.

**Comment by hashbangCoder** [comment]
Nevermind. I interpreted the paper (i was reading) completely wrong. Got a test case working now. Thanks anyway!

**Comment by suryabhupa** [comment]
To follow up on this -- if I'm using rnn's LSTM implementation, will cudnn.convert() work to convert it?

Are there any workarounds other than using cudnn.LSTM?
